### PR TITLE
Exporter

### PR DESCRIPTION
**Pull request by CSVdB**

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -19,5 +19,5 @@ extra-deps:
   subdirs:
     - typed-uuid
     - genvalidity-typed-uuid
-
+- QuickCheck-2.10.1
```
Comment (NorfairKing) [inline stack.yaml]: No, we already have a version of `QuickCheck`.

File: wolf-cub/src/Wolf/Cub/Draw.hs [hunk 2]
```diff
@@ -80,7 +80,8 @@ personEntryWidget now pe = go $ personEntryProperties pe
             , txt "   ("
             , withAttr
                   entryLastUpdatedAttr
-                  (txt (formatMomentNicely
+                  (txt
+                       (formatMomentNicely
```
Comment (NorfairKing) [inline wolf-cub/src/Wolf/Cub/Draw.hs]: No cosmetic changes please.

File: wolf-data/package.yaml [hunk 3]
```diff
@@ -24,6 +24,7 @@ dependencies:
 - path >=0.6 && <0.7
 - path-io >=1.3 && <1.4
 - process >=1.6 && <1.7
+- QuickCheck >=2.09 && <2.11
```
Comment (NorfairKing) [inline wolf-data/package.yaml]: Definitely not! No testing dependencies for the data packages.

File: wolf-data/src/Import.hs [hunk 4]
```diff
@@ -33,6 +33,7 @@ import Data.ByteString as X (ByteString)
 import Data.Map as X (Map)
 import Data.Text as X (Text)
 
+import Data.GenValidity as X
```
Comment (NorfairKing) [inline wolf-data/src/Import.hs]: See above.

File: wolf-data/src/Wolf/Data/Export/Result.hs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+instance (Show a, Show w, Show e) => Show (Result w e a) where
+    show (Success a) = "Success " ++ show a
+    show (Warning w a) = "Warning " ++ show w ++ " with value " ++ show a
+    show (Error e) = "Error " ++ show e
+
+instance (Validity a, Validity w, Validity e) => Validity (Result w e a) where
+    validate (Success a) = a <?!> "success"
+    validate (Warning w a) =
+        mconcat [w <?!> "warning in warning", a <?!> "value in warning"]
+    validate (Error e) = e <?!> "error"
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Result.hs]: If you `Derive Generic` for `Result`, this implementation will be generated by leaving off the part starting with `where`

File: wolf-data/src/Wolf/Data/Export/Result.hs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+instance (Eq a, Eq w, Eq e) => Eq (Result w e a) where
+    (==) (Success a) (Success b) = a == b
+    (==) (Warning w a) (Warning v b) = a == b && w == v
+    (==) (Error e) (Error e2) = e == e2
+    (==) _ _ = False
+
+instance (Show a, Show w, Show e) => Show (Result w e a) where
+    show (Success a) = "Success " ++ show a
+    show (Warning w a) = "Warning " ++ show w ++ " with value " ++ show a
+    show (Error e) = "Error " ++ show e
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Result.hs]: Derive it, don't write it.

File: wolf-data/src/Wolf/Data/Export/Result.hs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            Success x -> Warning w x
+            Warning w2 x -> Warning (mappend w w2) x
+            Error e -> Error e
+    (>>=) (Error e) _ = Error e
+
+instance (Eq a, Eq w, Eq e) => Eq (Result w e a) where
+    (==) (Success a) (Success b) = a == b
+    (==) (Warning w a) (Warning v b) = a == b && w == v
+    (==) (Error e) (Error e2) = e == e2
+    (==) _ _ = False
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Result.hs]: Derive it, don't write it.

File: wolf-data/src/Wolf/Data/Export/Result.hs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    isValid = isValidByValidating
+
+instance (GenUnchecked a, GenUnchecked w, GenUnchecked e) =>
+         GenUnchecked (Result w e a) where
+    genUnchecked =
+        oneof
+            [ Success <$> genUnchecked
+            , Warning <$> genUnchecked <*> genUnchecked
+            , Error <$> genUnchecked
+            ]
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Result.hs]: This can be generated.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 9]
```diff
@@ -89,3 +93,11 @@ instance ToJSON Repo where
             , "notes" .= repoNotes
             , "suggestions" .= repoSuggestions
             ]
+
+type Warning = String
+
+type Error = String
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: No. Use an ADT.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 10]
```diff
@@ -89,3 +93,11 @@ instance ToJSON Repo where
             , "notes" .= repoNotes
             , "suggestions" .= repoSuggestions
             ]
+
+type Warning = String
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: No. Use an ADT.

File: wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs [hunk 11]
```diff
@@ -15,3 +15,6 @@ spec = do
     eqSpec @Repo
     genValidSpec @Repo
     jsonSpecOnValid @Repo
+    functorSpecOnValid @ResultS
+    applicativeSpecOnValid @ResultS
+    monadSpecOnValid @ResultS
```
Comment (NorfairKing) [inline wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs:21]: COOL HUH!

**Review by NorfairKing — CHANGES_REQUESTED** [review]

File: wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs (hunk 11 above)
Comment (CSVdB) [reply wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs:21]: Yes!

Comment (CSVdB) [reply wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs:21]: Now you need to implement it such that this test is automatically generated, and that in addition he asks me "do you want to have an Eq instance for Result w e a? If so, here is my suggestion: [..]", and that I can just press 'y' and bam, default implementation. As in, a system which can
1. figure out that there is a default implementation for Eq and Show for this new type
2. can generate the code which generates the default implementation
3. automatically suggests it

**Comment by CSVdB** [comment]
Oops, I don't think I was meant to rebase this onto development yet... I'll also create a pull request onto development, is that's easier to use.

**Pull request closed without merging**